### ISSUE TITLE
stock quantity history valuation deathly slow - performance

### ISSUE THREAD

**Issue by gdgellatly** [odoo/odoo#29871]
Impacted versions:

Steps to reproduce:
I tried to reproduce on runbot but runbot kept breaking trying to setup.
Scenario is 
~90,000 products, ~2600 templates and say 100 or less stock moves total over 9 products with associated account moves. Mix of FIFO, FEFO and AVCO realtime valuation
Under Inventory Reporting - click Reporting -> Inventory Valuation, accept the defaults and Retrieve.
(Set timelimits higher first)

Current behavior:

Relevant 2 lines from log. Localhost dev test environment, no customisations affecting stock, no other activity.
```
2019-01-03 04:10:58,171 1 DEBUG master odoo.api: call stock.quantity.history(6,).open_table() 
2019-01-03 04:21:46,724 1 INFO master werkzeug: 192.168.32.11 - - [03/Jan/2019 04:21:46] "POST /web/dataset/call_button HTTP/1.1" 200 - 338259 338.038 310.524
```

I tried to use the profiler but my timeout was only set at 1 hour so it terminated too. I'm picking it is spending all its time in _compute_stock_value or maybe in qty_available.

Expected behavior:
Not 11 minutes.

Video/Screenshot link (optional):

**Comment by gdgellatly** [comment]
After taking a look it appears the issue is in _run_fifo_vacuum. More to come

**Comment by gdgellatly** [comment]
https://github.com/odoo/odoo/blob/ea301bef933b06995696a50c8f25aa022736321c/addons/stock_account/models/stock.py#L481

is there any reason why we can't just do this?
```python
    @api.model
    def _run_fifo_vacuum(self):
        # Call `_fifo_vacuum` on concerned moves
        fifo_valued_products = self.env['product.product']
        fifo_valued_products |= self.env['product.template'].search([('property_cost_method', '=', 'fifo')]).mapped(
            'product_variant_ids')
        fifo_valued_categories = self.env['product.category'].search([('property_cost_method', '=', 'fifo')])
        fifo_valued_products |= self.env['product.product'].search([('categ_id', 'child_of', fifo_valued_categories.ids)])
        moves_to_vacuum |= self.search(
                [('product_id', 'in', fifo_valued_products.ids), ('remaining_qty', '<', 0)] + self._get_all_base_domain())
        moves_to_vacuum._fifo_vacuum()
```
Currently we iterate over every fifo valued product to add the moves to vacuum, this triggers a record rule search and LOTS of queries. 5 or 6 for every fifo valued product.

**Comment by Yenthe666** [comment]
Hmm, actually, I think we could do what you suggest and would save a lot of queries.
@jco-odoo can you please check this?

**Comment by obayit** [comment]
Is there a way to implement this fix for version 14.0?
